accepted/wasmtime-debugging.md

```diff
@@ -0,0 +1,493 @@
+prerequisite for most other live-debugging features, as their behavior will be
+reused or extended. Once breakpoints have been implemented, the implementation
+of watch points, and step into can be done in parallel. As examining program
+state doesn’t require the ability to manipulate control flow while debugging, it
+can be built entirely off of the work required to support Post Mortem Debugging.
+
+Finally, the implementation of time-travel, record-and-replay, and interpreted
+(e.g. JS inside SpiderMonkey) debugging is left largely unspecified, as we will
+likely need to reevaluate their designs once we’ve finished the basic
+implementation of live debugging in Wasmtime.
+
+```mermaid
+flowchart
+    coreDumps(Core Dumps)
+    winch(Winch)
+    dap(DAP)
+    nullAdapter(Null Adapter)
+    dwarfAdapter(Dwarf Adapter)
+    debugAdapter(Debug Adapter Interface)
+    postMortemDebugging[Post Mortem Debugging]
+    breakPoints(Break Points)
+    watchPoints(Watch Points)
+    stepping(Steping)
+    examining(Examining Program State)
+    live[Live Debugging]
+    recordReplay(Record-and-replay)
+    spidermonkeyAdapter(SpiderMonkey Adapter)
+    timeTravel(Time-travel)
+
+    coreDumps --> dap
+    debugAdapter --> nullAdapter
+    debugAdapter --> dwarfAdapter
+    winch --> postMortemDebugging
+    dap --> postMortemDebugging
+    nullAdapter --> postMortemDebugging
+    dwarfAdapter --> postMortemDebugging
+    postMortemDebugging --> examining
+    postMortemDebugging --> breakPoints
+    winch --> breakPoints
+    breakPoints --> watchPoints
+    breakPoints --> stepping
+    examining --> live
+    watchPoints --> live
+    stepping --> live
+    live --> recordReplay
+    live --> spidermonkeyAdapter
+    live --> timeTravel
+```
+
+# Rationale and Alternatives
+
+## GDB or LLDB
+
+We could proceed instead by building workflows around an existing debugger
+that’s attached to Wasmtime itself. This would facilitate debugging at the Wasm
+level and the level of compiled modules, but would fall a little short when it
+comes to facilitating debugging interpreted languages. It would also require
+that users interact with the debugger using gdb’s native interface or through a
+Debug Adapter Protocol interface built on top of gdb requiring additional setup
+and configuration. We would need to build additional abstractions on top of gdb
+to support the same sort of interface that we’ve described for supporting
+interacting with SpiderMonkey’s debugger, and at that point we would be
+complicating our ability to support time-travel debugging or record+replay in
+the future by switching from debugging Wasm to debugging native code.
+
+Choosing an existing debugger like gdb or lldb would compromise our three
+principles outlined in the [motivation](#motivation) section &mdash;
+particularly principle (2) and supporting interpreted debuggees[^possible]
+&mdash; and therefore we feel that they are not viable choices.
```

Comment:
If someone out there really likes gdb or lldb, I'm pretty sure there's something that bridges DAP to the gdbserver protocol, making it possible for them to use their favorite debugger.  (Also worth noting that gdb already supports DAP from the get go, but only as a replacement to the gdbserver or its terminal UI.)

Reply:
> Also worth noting that gdb already supports DAP from the get go, but only as a replacement to the gdbserver or its terminal UI.

I would expect it to be only as replacement to it's terminal UI. Gdbserver provides low level access to the tracee like read/write memory or registers and set breakpoint at given address, then gdb on top of that handles debuginfo, calling functions, stepping whole lines/functions at a time, set breakpoint at given function or line, ... And finally the terminal UI, gdb-mi or DAP provide an interface with which you can interact with gdb.